File: stvs/src/main/java/edu/kit/iti/formal/stvs/model/table/HybridSpecification.java
package edu.kit.iti.formal.stvs.model.table;

import edu.kit.iti.formal.stvs.model.common.FreeVariableList;
import edu.kit.iti.formal.stvs.model.common.NullableProperty;
import edu.kit.iti.formal.stvs.model.common.Selection;
import edu.kit.iti.formal.stvs.model.common.ValidIoVariable;
import edu.kit.iti.formal.stvs.view.spec.timingdiagram.TimingDiagramCollectionController;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import javafx.collections.FXCollections;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;

/**
 * A {@link ConstraintSpecification} which also has an associated counterexample (a
 * {@link ConcreteSpecification}), a concrete instance ({@link ConcreteSpecification}) or both. This
 * class is the model on which the {@link edu.kit.iti.formal.stvs.view.spec.SpecificationController}
 * and {@link TimingDiagramCollectionController} operate.
 *
 * @author Benjamin Alt
 */
public class HybridSpecification extends ConstraintSpecification {

  private final NullableProperty<ConcreteSpecification> counterExample;
  private final NullableProperty<ConcreteSpecification> concreteInstance;
  private final boolean editable;
  private final ObservableList<HybridRow> rowsAsHybrid;

  /**
   * Stores which cell in the table is currently highlighted in the view. This is saved here in
   * order to allow the table and timing diagram to share the same reference.
   */
  private Selection selection;

  /**
   * Create a new, empty hybrid specification with a default name from a list of free variables.
   *
   * @param freeVariableList A list of initial free variables
   * @param editable True if this hybridSpecification is editable, false otherwise
   */
  public HybridSpecification(FreeVariableList freeVariableList, boolean editable) {
    this(DEFAULT_NAME, freeVariableList, editable);
  }

  /**
   * Create a new, empty hybrid specification with a given name and a list of free variables.
   *
   * @param name The name of the HybridSpecification
   * @param freeVariableList A list of initial free variables
   * @param editable True if this HybridSpecification is editable, false otherwise
   */
  public HybridSpecification(String name, FreeVariableList freeVariableList, boolean editable) {
    super(name, freeVariableList);
    this.editable = editable;
    this.selection = new Selection();
    this.counterExample = new NullableProperty<>();
    this.concreteInstance = new NullableProperty<>();
    this.rowsAsHybrid = FXCollections.observableArrayList();
    rowsAsHybrid.addListener(this::onHybridRowChanged);
    counterExample.addListener(event -> onCounterExampleChanged());
  }

  /**
   * Create a HybridSpecification from a {@link ConstraintSpecification}.
   *
   * @param sourceSpec The original {@link ConstraintSpecification}
   * @param editable True if this HybridSpecification is editable, false otherwise
   */
  public HybridSpecification(ConstraintSpecification sourceSpec, boolean editable) {
    this(sourceSpec.getName(), sourceSpec.getFreeVariableList(), editable);
    getColumnHeaders().addAll(sourceSpec.getColumnHeaders());

    for (int rowIndex = 0; rowIndex < sourceSpec.getRows().size(); rowIndex++) {
      HybridRow row = new HybridRow(sourceSpec.getRows().get(rowIndex),
          sourceSpec.getDurations().get(rowIndex));
      row.updateCounterExampleCells(rowIndex, getCounterExample());
      rowsAsHybrid.add(row);
    }
  }

  private void onCounterExampleChanged() {
    for (int rowIndex = 0; rowIndex < getRows().size(); rowIndex++) {
      rowsAsHybrid.get(rowIndex).updateCounterExampleCells(rowIndex, getCounterExample());
    }
  }

  private void onHybridRowChanged(ListChangeListener.Change<? extends HybridRow> change) {
    while (change.next()) {
      if (change.wasAdded()) {
        List<SpecificationRow<ConstraintCell>> rowsToBeAdded = new ArrayList<>();
        List<ConstraintDuration> durationsToBeAdded = new ArrayList<>();
        for (HybridRow row : change.getAddedSubList()) {
          SpecificationRow<ConstraintCell> rowToBeAdded = row.getSourceRow();
          rowToBeAdded.commentProperty().bindBidirectional(row.commentProperty());
          rowsToBeAdded.add(rowToBeAdded);
          durationsToBeAdded.add(row.getDuration().getCell());
        }
        getRows().addAll(change.getFrom(), rowsToBeAdded);
        getDurations().addAll(change.getFrom(), durationsToBeAdded);
      }
      if (change.wasRemoved()) {
        getRows().remove(change.getFrom(), change.getFrom() + change.getRemovedSize());
        getDurations().remove(change.getFrom(), change.getFrom() + change.getRemovedSize());
      }
    }
  }

  public Optional<ConcreteSpecification> getCounterExample() {
    return Optional.ofNullable(counterExample.get());
  }

  /**
   * Sets the counterexample for this hybrid specification. This will automatically update
   * the {@link HybridRow}'s counterexample cells in {@link #getHybridRows()} from the given
   * counterexample.
   *
   * @param counterExample the concrete specification to be shown in-place in the gui
   * @throws IllegalArgumentException if the given concrete instance's column headers don't match
   *                                  this specification's column headers
   */
  public void setCounterExample(ConcreteSpecification counterExample) {
    if (counterExample != null) {
      if (!columnHeadersMatch(counterExample.columnHeaders)) {
        throw new IllegalArgumentException("The column headers of the concrete instance are not "
            + "compatible with this hybrid specification");
      }
      this.counterExample.set(counterExample);
    }
  }

  public ObservableList<HybridRow> getHybridRows() {
    return rowsAsHybrid;
  }

  public Selection getSelection() {
    return selection;
  }

  public boolean isEditable() {
    return editable;
  }

  public Optional<ConcreteSpecification> getConcreteInstance() {
    return Optional.ofNullable(concreteInstance.get());
  }

  public NullableProperty<ConcreteSpecification> concreteInstanceProperty() {
    return concreteInstance;
  }

  public NullableProperty<ConcreteSpecification> counterExampleProperty() {
    return counterExample;
  }

  /**
   * <p>Set the generated concrete instance for this hybrid specification, that is the
   * concretized constraint specification. This is concrete instance is then used from
   * the {@link TimingDiagramCollectionController} to view a timing diagram.</p>
   *
   * @param concreteInstance the concretized constraint specification
   * @throws IllegalArgumentException if the given concrete instance's column headers don't match
   *                                  this specification's column headers
   */
  public void setConcreteInstance(ConcreteSpecification concreteInstance) {
    if (concreteInstance != null) {
      if (!columnHeadersMatch(concreteInstance.columnHeaders)) {
        throw new IllegalArgumentException("The column headers of the concrete instance are not "
            + "compatible with this hybrid specification");
      }
      this.concreteInstance.set(concreteInstance);
    }
  }

  private boolean columnHeadersMatch(ObservableList<ValidIoVariable> columnHeaders) {
    if (this.columnHeaders.size() != columnHeaders.size()) {
      return false;
    }
    for (int i = 0; i < this.columnHeaders.size(); i++) {
      if (!this.columnHeaders.get(i).matches(columnHeaders.get(i))) {
        return false;
      }
    }
    return true;
  }

  public void removeConcreteInstance() {
    concreteInstance.set(null);
  }

  @Override
  public int hashCode() {
    int result = super.hashCode();
    result = 31 * result + (getCounterExample() != null ? getCounterExample().hashCode() : 0);
    result = 31 * result + (getConcreteInstance() != null ? getConcreteInstance().hashCode() : 0);
    result = 31 * result + (isEditable() ? 1 : 0);
    result = 31 * result + (rowsAsHybrid != null ? rowsAsHybrid.hashCode() : 0);
    result = 31 * result + (getSelection() != null ? getSelection().hashCode() : 0);
    return result;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null || getClass() != obj.getClass()) {
      return false;
    }
    if (!super.equals(obj)) {
      return false;
    }

    HybridSpecification that = (HybridSpecification) obj;

    if (isEditable() != that.isEditable()) {
      return false;
    }
    if (getCounterExample() != null ? !getCounterExample().equals(that.getCounterExample())
        : that.getCounterExample() != null) {
      return false;
    }
    if (getConcreteInstance() != null ? !getConcreteInstance().equals(that.getConcreteInstance())
        : that.getConcreteInstance() != null) {
      return false;
    }
    if (rowsAsHybrid != null ? !rowsAsHybrid.equals(that.rowsAsHybrid)
        : that.rowsAsHybrid != null) {
      return false;
    }
    return getSelection() != null ? getSelection().equals(that.getSelection())
        : that.getSelection() == null;
  }
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/model/table/HybridSpecification.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.table

import edu.kit.iti.formal.stvs.model.common.FreeVariableList
import edu.kit.iti.formal.stvs.model.common.NullableProperty
import edu.kit.iti.formal.stvs.model.common.Selection
import edu.kit.iti.formal.stvs.model.common.ValidIoVariable
import javafx.beans.Observable
import javafx.beans.property.SimpleListProperty
import javafx.collections.FXCollections
import javafx.collections.ListChangeListener
import javafx.collections.ObservableList
import tornadofx.getValue
import tornadofx.setValue

/**
 * A [ConstraintSpecification] which also has an associated counterexample (a
 * [ConcreteSpecification]), a concrete instance ([ConcreteSpecification]) or both. This
 * class is the model on which the [edu.kit.iti.formal.stvs.view.spec.SpecificationController]
 * and [TimingDiagramCollectionController] operate.
 *
 * @author Benjamin Alt
 */
class HybridSpecification(name: String, freeVariableList: FreeVariableList, val isEditable: Boolean) : ConstraintSpecification(name, freeVariableList) {
    val counterExampleProperty = NullableProperty<ConcreteSpecification?>()
    var counterExample: ConcreteSpecification?
        get() = counterExampleProperty.get()
        /**
         * Sets the counterexample for this hybrid specification. This will automatically update
         * the [HybridRow]'s counterexample cells in [.getHybridRows] from the given
         * counterexample.
         *
         * @param counterExample the concrete specification to be shown in-place in the gui
         * @throws IllegalArgumentException if the given concrete instance's column headers don't match
         * this specification's column headers
         */
        set(value) {
            if (value != null) {
                require(columnHeadersMatch(value.columnHeaders)) {
                    (
                        "The column headers of the concrete instance are not " +
                            "compatible with this hybrid specification"
                        )
                }
            }
            this.counterExampleProperty.set(value)
        }

    val concreteInstanceProperty = NullableProperty<ConcreteSpecification?>()
    var concreteInstance: ConcreteSpecification?
        get() = concreteInstanceProperty.get()
        /**
         *
         * Set the generated concrete instance for this hybrid specification, that is the
         * concretized constraint specification. This is concrete instance is then used from
         * the [TimingDiagramCollectionController] to view a timing diagram.
         *
         * @param concreteInstance the concretized constraint specification
         * @throws IllegalArgumentException if the given concrete instance's column headers don't match
         * this specification's column headers
         */
        set(value) {
            if (value != null) {
                require(columnHeadersMatch(value.columnHeaders)) {
                    (
                        "The column headers of the concrete instance are not " +
                            "compatible with this hybrid specification"
                        )
                }
            }
            this.concreteInstanceProperty.set(value)
        }

    val hybridRowsProperty = SimpleListProperty<HybridRow>(FXCollections.observableArrayList())
    var hybridRows: ObservableList<HybridRow> by hybridRowsProperty

    /**
     * Stores which cell in the table is currently highlighted in the view. This is saved here in
     * order to allow the table and timing diagram to share the same reference.
     */
    private val selection = Selection()

    /**
     * Create a new, empty hybrid specification with a default name from a list of free variables.
     *
     * @param freeVariableList A list of initial free variables
     * @param editable True if this hybridSpecification is editable, false otherwise
     */
    constructor(
        freeVariableList: FreeVariableList,
        editable: Boolean,
    ) : this(DEFAULT_NAME, freeVariableList, editable)

    /**
     * Create a new, empty hybrid specification with a given name and a list of free variables.
     */
    init {
        hybridRowsProperty.addListener(
            ListChangeListener<HybridRow> { change ->
                this.onHybridRowChanged(change)
            },
        )
        counterExampleProperty.addListener { event: Observable? -> onCounterExampleChanged() }
    }

    /**
     * Create a HybridSpecification from a [ConstraintSpecification].
     *
     * @param sourceSpec The original [ConstraintSpecification]
     * @param editable True if this HybridSpecification is editable, false otherwise
     */
    constructor(sourceSpec: ConstraintSpecification, editable: Boolean) : this(
        sourceSpec.name,
        sourceSpec.freeVariableList,
        editable,
    ) {
        columnHeaders.addAll(sourceSpec.columnHeaders)

        for (rowIndex in sourceSpec.rows.indices) {
            val row = HybridRow(
                sourceSpec.rows[rowIndex],
                sourceSpec.durations[rowIndex],
            )
            row.updateCounterExampleCells(rowIndex, counterExample)
            hybridRowsProperty.add(row)
        }
    }

    private fun onCounterExampleChanged() {
        for (rowIndex in rows.indices) {
            hybridRowsProperty[rowIndex]!!.updateCounterExampleCells(rowIndex, counterExample)
        }
    }

    private fun onHybridRowChanged(change: ListChangeListener.Change<out HybridRow>) {
        while (change.next()) {
            if (change.wasAdded()) {
                val rowsToBeAdded = arrayListOf<SpecificationRow<ConstraintCell>>()
                val durationsToBeAdded: MutableList<ConstraintDuration?> = ArrayList()
                for (row in change.addedSubList) {
                    val rowToBeAdded = row.sourceRow
                    rowToBeAdded.commentProperty.bindBidirectional(row!!.commentProperty)
                    rowsToBeAdded.add(rowToBeAdded)
                    durationsToBeAdded.add(row.duration.cell!!)
                }
                rows.addAll(change.from, rowsToBeAdded)
                durations.addAll(change.from, durationsToBeAdded)
            }
            if (change.wasRemoved()) {
                rows.remove(change.from, change.from + change.removedSize)
                durations.remove(change.from, change.from + change.removedSize)
            }
        }
    }

    fun getSelection(): Selection = selection

    private fun columnHeadersMatch(columnHeaders: ObservableList<ValidIoVariable>): Boolean {
        if (this.columnHeaders.size != columnHeaders.size) {
            return false
        }
        for (i in this.columnHeaders.indices) {
            if (!this.columnHeaders[i]!!.matches(columnHeaders[i]!!)) {
                return false
            }
        }
        return true
    }

    fun removeConcreteInstance() {
        concreteInstanceProperty.set(null)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        if (!super.equals(other)) return false

        other as HybridSpecification

        if (isEditable != other.isEditable) return false
        if (counterExample != other.counterExample) return false
        if (concreteInstance != other.concreteInstance) return false
        if (hybridRows != other.hybridRows) return false
        if (selection != other.selection) return false
        if (columnHeaders != other.columnHeaders) return false
        if (comment != other.comment) return false
        if (rows != other.rows) return false
        if (freeVariableList != other.freeVariableList) return false
        if (durations != other.durations) return false
        if (name != other.name) return false
        if (concreteInstance != other.concreteInstance) return false
        if (counterExample != other.counterExample) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + isEditable.hashCode()
        result = 31 * result + (counterExample?.hashCode() ?: 0)
        result = 31 * result + (concreteInstance?.hashCode() ?: 0)
        result = 31 * result + hybridRows.hashCode()
        result = 31 * result + selection.hashCode()
        result = 31 * result + columnHeaders.hashCode()
        result = 31 * result + rows.hashCode()
        result = 31 * result + freeVariableList.hashCode()
        result = 31 * result + durations.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + counterExample.hashCode()
        result = 31 * result + concreteInstance.hashCode()
        return result
    }
}